EIPS/abstraction.md

```diff
@@ -0,0 +1,94 @@
+
+    EIP: <to be assigned>
+    Title: Abstraction of transaction origin and signature
+    Author: Vitalik Buterin
+    Type: Standard Track
+    Category: Core
+    Status: Draft
+    Created: 2017-02-10
+    
+### Summary
+
+Implements a set of changes that serve the combined purpose of "abstracting out" signature verification and nonce checking, allowing users to create "account contracts" that perform any desired signature/nonce checks instead of using the mechanism that is currently hard-coded into transaction processing.
+
+### Parameters
+
+* METROPOLIS_FORK_BLKNUM: TBD
+* CHAIN_ID: same as used for EIP 155 (ie. 1 for mainnet, 3 for testnet)
+* NULL_SENDER: 2**160 - 1
+
+### Specification
+
+If `block.number >= METROPOLIS_FORK_BLKNUM`, then:
+1. If the signature of a transaction is `(CHAIN_ID, 0, 0)` (ie. `r = s = 0`, `v = CHAIN_ID`), then treat it as valid and set the sender address to `NULL_SENDER`
+2. Transactions of this form MUST have gasprice = 0, nonce = 0, value = 0, and do NOT increment the nonce of account NULL_SENDER.
+3. Create a new opcode at `0xfb`, `CREATE2`, with 4 stack arguments (value, salt, mem_start, mem_size) which sets the creation address to `sha3(sender + salt + sha3(init code)) % 2**160`, where `salt` is always represented as a 32-byte value.
+4. Add to _all_ contract creation operations, including transactions and opcodes, the rule that if a contract at that address already exists and has non-empty code OR non-empty nonce, the operation fails and returns 0 as if the init code had run out of gas. If an account has empty code and nonce but nonempty balance, the creation operation may still succeed.
+
+### Rationale
+
+The goal of these changes is to set the stage for abstraction of account security. Instead of having an in-protocol mechanism where ECDSA and the default nonce scheme are enshrined as the only "standard" way to secure an account, we take initial steps toward a model where in the long term all accounts are contracts, contracts can pay for gas, and users are free to define their own security model.
+
+Under EIP 86, we can expect users to store their ether in contracts, whose code might look like the following (example in Serpent):
+
+```python
+# Get signature from tx data
+sig_v = ~calldataload(0)
+sig_r = ~calldataload(32)
+sig_s = ~calldataload(64)
+# Get tx arguments
+tx_nonce = ~calldataload(96)
+tx_to = ~calldataload(128)
+tx_value = ~calldataload(160)
+tx_gasprice = ~calldataload(192)
+tx_data = string(~calldatasize() - 224)
+~calldataload(tx_data, 224, ~calldatasize())
+# Get signing hash
+signing_data = string(~calldatasize() - 64)
+~mstore(signing_data, tx.startgas)
+~calldataload(signing_data + 32, 96, ~calldatasize() - 96)
+signing_hash = sha3(signing_data:str)
+# Perform usual checks
+prev_nonce = ~sload(-1)
+assert tx_nonce == prev_nonce + 1
+assert self.balance >= tx_value + tx_gasprice * tx.startgas
+assert ~ecrecover(signing_hash, sig_v, sig_r, sig_s) == <pubkey hash here>
```

Comment:
Should we also abstract the pubkey verification? It could even be abstracted internally, by having a number of precompile contracts where you can call one with an opcode sigVerify. The opcode could even have a default option (where you can call it without any argument).